src/bundle/Security/TwoFactor/Provider/TwoFactorProviderDeciderInterface.php

```diff
@@ -0,0 +1,17 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Scheb\TwoFactorBundle\Security\TwoFactor\Provider;
+
+use Scheb\TwoFactorBundle\Security\Authentication\Token\TwoFactorTokenInterface;
+
+interface TwoFactorProviderDeciderInterface
+{
+    /**
+     * Return the alias of the preferred two-factor provider.
+     *
+     * @param string[] $activeProviders
+     */
+    public function getPreferredTwoFactorProvider(array $activeProviders, TwoFactorTokenInterface $token, object $user): string|null;
```

Comment:
I'd suggest to adjust the interface slightly. Instead of passing the $user, better pass a `AuthenticationContextInterface $context` here, which provides pretty much all relevant information to make a decision about the preferred provider.